### DISCUSSION TITLE
Does anyone use LIBRARIES here? Which ones do and don't work?

### DISCUSSION THREAD

**Discussion by SpenceKonde** [SpenceKonde/megaTinyCore#351] [Library discussion]
Please help me out here. People keep asking me what library do I use for this? what library do I use for that? I don't know! I spend all my time maintaining the core, not using it! 

If people who are using this core could let me know which libraries they have used (include link to where that library comes from) or tried to use and whether they did or did not work, so I can gather all that together on a page with a big table of libraries, that'd be great.

Thanks!

**Comment by SpenceKonde** [comment]
Lot of people have been asking about I2C device libraries, particularly screens, someone asked about GPS (GPS?! I thought any old GPS library should work, it's just serial!) And so on....

**Reply by iansexton** [reply]
The Adafruit MLX90614 (I2C non contact temperature) seems to work nicely :)
https://learn.adafruit.com/using-melexis-mlx90614-non-contact-sensors/wiring-and-test

**Reply by iansexton** [reply]
Adafruit VEML6070 (I2C UV light sensor) seems to work nicely.

**Reply by iansexton** [reply]
Adafruit BME280 doesn't compile:
Arduino: 1.8.13 (Windows 10), Board: "ATtiny1614/1604/814/804/414/404/214/204 (Optiboot), ATtiny1614, 20 MHz Internal, Enabled (default timer), 1.8V (5 MHz or less), Disabled/Disabled, Closer to 5v, No (saves space), TX:5 (PB2), RX:4 (PB3), UPDI (powercycle to enter bootloader), 8ms"

C:\Users\Ian\Dropbox\arduino_esp\libraries\Adafruit_Sensor\Adafruit_Sensor.cpp:2:10: fatal error: pgmspace.h: No such file or directory

 #include <pgmspace.h>

          ^~~~~~~~~~~~

compilation terminated.

exit status 1

Error compiling for board ATtiny1614/1604/814/804/414/404/214/204 (Optiboot).

**Reply by iansexton** [reply]
Adafruit_BMP085_U has the same problem

**Comment by SpenceKonde** [comment]
Update the Adafruit Sensor library. 
you are using a copy that dates to ancient times (and wouldn't compile anywhere, since the file is located in avr/pgmspace.) Reference to it was removed from the library in 2015 , 
https://github.com/adafruit/Adafruit_Sensor/commit/04a0f643903f54e7d9c69aaedd0fe1f5b2a31c4d

**Reply by iansexton** [reply]
OK Works fine compiling for ESP8266

**Comment by SpenceKonde** [comment]
AAHA! In my brief search during which I discovered that, I saw someone mentioning that pgmspace was in a different location on the ESP8266. That sounds like an ESP8266-only version of the library...

**Reply by iansexton** [reply]
Such fun :)
You could well be right!

**Reply by iansexton** [reply]
Pololu VL53L0X works nicely

**Reply by iansexton** [reply]
Suggest you make a table of 'tested' libraries
I had problems with I2C LCD & was pointed to a library which compiles to 5K! When I was a boy it took a few lines of assembly to drive these things :(

**Reply by iansexton** [reply]
Guess what - recent Adafruit BMP180 & BME280 libraries both work :)
Although I'd forgotten how big these are...

**Reply by SpenceKonde** [reply]
A table of tested libraries? YES THAT IS EXACTLY WHAT WE ARE DOING!!

**Comment by SpenceKonde** [comment]
Well, I've put the page here - but it's pretty dissappointing considering how little information has been collected, and how I've only got one that is unambiguously identified with a link... https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/LibraryCompatibility.md

**Reply by iansexton** [reply]
You missed another two from my comment above.
Adafruit BME280 & BMP180 both work with current libraries.

**Reply by iansexton** [reply]
How do we find your table without visiting this page? It's not obvious to me :(

**Reply by SpenceKonde** [reply]
By clicking the link I have not yet added to the readme, I will add it when I add these new entries, but I have to finish the day job, pack and ship half a dozen orders and put out a bunch of other fires first.

**Comment by sideswiper22** [comment]
Confirmed and tested working with the LiquidCrystal_I2C Library. Most I2C libraries for the Arduino Uno seems to be compatible with this core. The only places where I did have problems with libraries are when I have to interface via SPI.

**Reply by SpenceKonde** [reply]
LOL, LiquidCrystal_I2C is the worst offender of all time in terms of "different library with same name"... I have two versions that are identically named with COMPLETELY DIFFERENT INTERFACES - no part of code written for one would compile with the other included! 

So without knowing which version you're talking about, this is not useful. 

Can you name any libraries that don't work? It would be extremely useful to be able to also include libraries that are known to not work, so others are fore-warned that they shouldn't waste time with them, unless they are up to the challenge of figureing out the problem (and. I hope, sharing their solution - maybe even being a hero and submitting a PR!)

**Reply by sideswiper22** [reply]
It's this version of the library that I tried and worked https://github.com/johnrickman/LiquidCrystal_I2C

**Reply by SpenceKonde** [reply]
> It's this version of the library that I tried and worked https://github.com/johnrickman/LiquidCrystal_I2C

Dear god, another one. We're now up to two reports of a working LiquidCrystal_I2C library.... BOTH OF WHICH ARE NO LONGER MAINTAINED BY THEIR AUTHORS and have been moved to alternate hosts or made read-only. (also, wtf is gitlab? He appears to have moved his open source library that others are using to a private repo on gitlab...?)

**Reply by SpenceKonde** [reply]
Ah, found the story in the issues. He moved to gitlab, then attached a BSD license to his derivative works that were previously licensed LGPL v2.1 (which is not consistent with the LGPL), someone called him on it, he initially seemed open to doing the correct thing, but then he either went emo or got spooked and deleted the gitlab account and stopped responding to any attempts at contact, despite some people puttnig in considerable effort to reach him.

So, we stll don't have a version of liquid crystal which I can recommend that people use :-(

**Comment by ldalep** [comment]
https://github.com/PaulStoffregen/Time/blob/master/TimeLib.h
https://github.com/adafruit/RTClib
Works with Atiny 1614

Is this the correct format for what you want?

**Reply by SpenceKonde** [reply]
That's great! thanks, will add.

**Reply by ldalep** [reply]
There is and I don't know if this is the correct place but the Softserial library allows the use of -1 to negate a TX or RX port.  Not sure if that is a breaker for this library but it would be nice to use and UNused TX or RX for another port.   Maybe just to hard to implement.

**Reply by SpenceKonde** [reply]
It should work, SoftwareSerial does that, and I made an excepton to the compiletme error checking to not throw an error over acompiletime known call to digitalRead /digitalWrite/pinMode on -1 (aka NOT_A_PIN or 255)

IIRC the only problem with that was that it blindly charged ahead

**Reply by ldalep** [reply]
Thanks- I did't know that it was fixed

**Reply by SpenceKonde** [reply]
I am actually considering changing it back and modifying SoftwareSerial to recognize that it's been passed a non-pin sooner; I wish I knew if anything else did that; it's pretty shoddy IMO to just merrily pass on NOT_A_PIN instead of testing for it being a valid pin....

**Comment by lou270** [comment]
I had a problem with OneWire, check this https://github.com/SpenceKonde/OneWire/pull/1

**Reply by SpenceKonde** [reply]
Why does the core *not* just use _delay_us for constant delays? Pretty sure it would be better (as long as the delay is a compile-time known constant, at least - I think that's required)

**Reply by SpenceKonde** [reply]
This will be done in 2.2.10 (if such a thing is released) and 2.3.0 (if not)

**Comment by lerouxb** [comment]
I've been using Tiny4kOLED: https://github.com/datacute/Tiny4kOLED

It is really nice in that it uses the extra memory in the displays based on the SSD1306 to implement double buffering on 128x32 displays.

And those displays are ridiculously cheap: https://www.aliexpress.com/wholesale?catId=0&initiative_id=SB_20210313120846&SearchText=SSD1306+%22128x32%22+0.91%22

**Reply by lerouxb** [reply]
Coincidentally this was the final straw in making me just go with Arduino rather than Atmel Start.

Atmel Start / from scratch: You get access to all the functionality these chips have and it is highly configurable.
Arduino: Libraries and a community. There's also value in being able to say your product is open source, based on arduino and people can modify it using the Arduino IDE.

When I got to the point where I was about to port Arduino libraries over to ASF I realised just how ridiculous it all is.

**Reply by grandaspanna** [reply]
I'm using a bit of a blend of Atmel Start and Arduino. Start is great for generating some of the device-specific code for different peripherals and then I just copy that code into an Arduino project. Helps with getting register settings etc setup properly.

**Reply by SpenceKonde** [reply]
@lerouxb - and this is working for the megaTinyCore parts, not ATTinyCore? Not that i question it, just that it looks like they refer specifically to a lot of classic-AVR tiny stuff  specifically mentioned. But if it works here, I would venture a guess that it doesn't need anything beyond a working wire library.

**Reply by lerouxb** [reply]
Yup. megaTinyCore. Using it with attiny3216. Definitely just a working wire lib.

**Comment by iansexton** [comment]
rc-switch seems to be working. https://github.com/sui77/rc-switch
I've only tried the transmit side at 20MHz clock. This used code from an old Tiny85 project & wasn't an exhaustive test. It worked for me, so maybe it will work for you too.

**Reply by SpenceKonde** [reply]
What part dd you test it with - Am surprised to hear that it worked, great news if so. 

Pretty sure rx will not work

**Reply by iansexton** [reply]
Using cheap h34a 433MHz ASK transmitter on a 1614. I've had another play - it seems to work at all clock rates on the 1614 which it didn't do on a Tiny 85.
My immediate use case is ringing wireless door bells so I'm not so bothered about rx not working. I sometimes dabble with Open MQTT Gateway which uses rc-switch so it's quite nice that it works :)
If I can find a receiver (and a spare hour) I'll have a look at the receive side.

**Reply by SpenceKonde** [reply]
Well that is surprising and great news!

**Comment by DanSkareda** [comment]
MegaTinyCore (AVR-1: Attiny1614 or Attiny3216) - this library works:

digital thermometer DS18B20 (oneWire) - https://github.com/pstolarz/OneWireNg

**Reply by DanSkareda** [reply]
I bought attiny1624 now. Please, can you update core about 2. series?

**Reply by SpenceKonde** [reply]
Thanks, I have that one entered on the other tab but not submitted tyo github yet. 

Github latest (manual install) works with the 2-series except that there is no analogRead on 2-series. Or it should work! I released working support weeks ago, but nobody has gotten back to me with any feedback about whether the parts of the core other than analogRead() are working correctly on 2-series, and I haven't had time to work on it. (my day job is software test. So I hate testing when I'm working on my own stuff... so I test in productiion... except my test team has been quiet as mice re: 2-series. Everyone asked about it, I put out what I could get ready quickly - and then? crickets!

**Reply by iansexton** [reply]
You might wait a while. 2 series parts are not readily available in the UK.
I have to buy 25x for the 1 series parts to make it sensible (RS Components)

**Reply by SpenceKonde** [reply]
Well, @DanSkareda could certainly start banging on the new version of the core and telling me what does and doesn't work!

**Reply by DanSkareda** [reply]
I manual installed megaTinyCore. Then I try compile this (almost) example sketch for Attiny1624 (I am sorry, my English not good):

void setup() {
  pinMode(1, OUTPUT);
}

// the loop function runs over and over again forever
void loop() {
  digitalWrite(1, HIGH);   // turn the LED on (HIGH is the voltage level)
  delay(1000);                       // wait for a second
  digitalWrite(1, LOW);    // turn the LED off by making the voltage LOW
  delay(1000);                       // wait for a second
}

Compile ended with this mistakes. I do not know where is problem.

In file included from sketch\Blink.ino.cpp:1:0:
C:\Users\Dan\Documents\Arduino\hardware\megaTinyCore-master\megaavr\cores\megatinycore/Arduino.h:401:8: error: macro names must be identifiers
 #ifndef(MEGATINYCORE)
        ^
exit status 1

**Reply by SpenceKonde** [reply]
aaaarrrghhh why isn't that fix in?! I swear I fixed that!

(see, this is why I need people to test this stuff - I thought everything was compiling just fine!

**Reply by SpenceKonde** [reply]
Thanks, @DanSkareda  things compile on master branch again (and several other changes have been brought from 2.2.x to 2.3.0

**Reply by DanSkareda** [reply]
Thank you very much. It is ok, now.

**Comment by SpenceKonde** [comment]
Argh, lost all my work on this!

**Comment by grandaspanna** [comment]
Just verified the Adafruit LED backpack library, at least for the 7-segment variant. Also worked with the Wire.swap(1) method to use the alternate i2c pins. Nice!
Adafruit-LED-Backpack-Library
Versions 1.1.8
https://github.com/adafruit/Adafruit_LED_Backpack
Tested on 1614

**Comment by grandaspanna** [comment]
The Adafruit TLC5947 driver works on my attiny412 board.
https://github.com/adafruit/Adafruit_TLC5947
![IMG_5667](https://user-images.githubusercontent.com/54263749/113805260-90383c00-97a3-11eb-92ea-d3f0bdb773b7.jpg)

**Reply by trimarco232** [reply]
Hi,
very nice board !
of course it works, but the adafruit library isn't a good way to handle tlc5947 with mcus like tiny412 and its ridiculous amount of ram
this library wastes 25% of the ram while using 16 bit words for 12 bit needed
in order to make it better, one rather put two 12 bits words into three bytes ...

**Reply by grandaspanna** [reply]
It's a fair point, but with a single controller and 24 channels, I think it's only 8 bytes I'm tossing aside. That's enough headroom for my current purposes. I've more often hit size problems dealing with OLED displays (either code/ram or both). For me, it's usually easy to go up to a larger microcontroller than waste too much time squeezing them dry. I appreciate not everyone has that luxury. I'm still very grateful for the work others put in to make these cores and libraries available.

**Reply by trimarco232** [reply]
yes, it's only 12 bytes (48 vs 36)

**Comment by grandaspanna** [comment]
Gave up trying to leverage the Microchip touch library, so implemented a standalone touch controller instead. Picked an i2c one that was readily available locally.

To this end, the CAP1296 library (https://github.com/mattThurstan/CAP1296) is working nicely on an attiny1614 breakout with 2.3.1

Two comments:

1.  The library has embedded Serial.print statements. I've disabled them in my local copy
2. The default i2c address is wrong, but easily sorted by passing 0x28 to the begin() method

![IMG_5789](https://user-images.githubusercontent.com/54263749/116763266-8df29600-aa60-11eb-882b-49ed1b08e603.jpeg)

**Reply by SpenceKonde** [reply]
The heck is the difference between the library and the one it's based on? They look nearly identical....

**Comment by rneurink** [comment]
Used a few I2C libraries and as expected they are working as expected:

MCP4725 compiles and works https://github.com/adafruit/Adafruit_MCP4725
MCP23017 compiles and works https://github.com/adafruit/Adafruit-MCP23017-Arduino-Library

Some self-advertising:
VL53L1X_ULD (Time of flight ultra-light driver for the VL53L1X sensor from ST, relatively mature) https://github.com/rneurink/VL53L1X_ULD

**Reply by SpenceKonde** [reply]
Great great, will add to page - somehow I wasn't getting notifications about this thread.

**Comment by mechatroniks-git** [comment]
This library works well with ATtiny, tested with 2-series. https://github.com/rlogiacco/CircularBuffer

**Reply by SpenceKonde** [reply]
Now that's a neat library. I mean, of course it works, that's no surprise considering what it is doing... but I gotta give the dude credit. He who wrote that library is versed in dark arts with which I dare not involve myself with...

And he keeps a version history too...

**Comment by dattasaurabh82** [comment]
- [x] [SparkFun_RV-8803_Arduino_Library](https://github.com/sparkfun/SparkFun_RV-8803_Arduino_Library) works fine but consumes a lot of memory. Tested on `ATTINY-1605`

**Reply by SpenceKonde** [reply]
Which kind of memory, ram or flash? Or both?

**Comment by chipperdoodles** [comment]
It's not in the release branch but: 

[https://github.com/FastLED/FastLED/pull/1183](https://github.com/FastLED/FastLED/pull/1183)


I tried out a branch with the pull merged and got it to compile and run on a 3217, just tried the analog example and it worked clocked at 1mhz. If using an addressable you'd need higher clock rates but was exciting to see work nontheless. I'll comment on this when i notice support entering official fastled release.

**Comment by SpenceKonde** [comment]
Well that['s good news! About goddamned time. I looked at that library thinking I would be able to add it and I quickly lost interest. IMHO that library exhibits extreme kitchensinkism.

**Comment by hmeijdam** [comment]
I have tested the [S-Port Library](https://github.com/RealTadango/FrSky) version 1.0.2
https://www.arduino.cc/reference//en/libraries/s.port-sensor-library-for-frsky/
I can confirm it's working with this core, 

However there is a caveat that bit me first!

I used the "simplesensor" example that comes with the library, which compiles just under 4K for my test Attiny416 but that failed to work. So I thought "not compatible".
A few days later I compiled the same example for 8K (Attiny806) and 16K  1614 and 1626) parts and it works fine.
So I suspect it's running out of SRAM and gets a stack/heap collision?
I am curious to learn what's causing this. Is it the use of classes?

Also it is using softwareserial to get the onewire inverted serial communication. These Tinies seem to have a built-in one wire open drain mode on the UART, where the internal pullup is used to drive the UART pin high. Can that be used for an inverted signal also?

**Reply by trimarco232** [reply]
Hi hmeijdam,
it should work, just try : set the ODME, PULLUPEN, and INVEN bits
(of course, the pull-up will not be inverted to pull-down)
